README.md
## SocialApp

It's the next big thing!  Actually it's pretty unimpressive.  But it does contain some
useful build automation scripts.

## Prerequisites

This build system requires the following:

- Ruby
- Bundler
- Rake

Run `bundle install` to get the remaining dependencies.

## Building the app

A simple Xcode build:

````
rake xcode:build
````

Running tests:

````
rake test
````

Building an IPA:

````
rake build_ipa
````

Generating documentation:

````
rake docs
````

## Continuous Integration Servers

On a ci server, you can run `rake ci` to automatically install provisioning profiles & 
certificates, build an ipa, zip dsyms, and generate documentation.

See the slides [here](https://speakerdeck.com/subdigital/ios-build-automation-with-jenkins).








RELEASE_NOTES
These are the release notes


scripts/test.rb
# based on https://gist.github.com/3349345
# Thanks, @alloy!
#
# To get your project ready for this, you'll have to create a scheme for your unit test project, make sure run is checked in 
# the build step, and then delete the Test Host setting.
# Also, make sure you have the colored and open4 gems installed.

require 'rubygems'
require 'colored'
require 'pathname'
require 'open4'

# Change these to what is appropriate for your project
source_root = File.expand_path("../", __FILE__)
workspace = "SocialApp.xcworkspace"
scheme = "SocialApp"

verbose = !!ARGV.delete("--verbose")


class TestOutput < Array
  def initialize(io, verbose, source_root)
    @io, @verbose, @source_root = io, verbose, Pathname.new(source_root)
  end

  def <<(line)
    return if !@verbose && line =~ /^Test (Case|Suite)/
    super
    @io << case line
    when /^Run test case/
      line.bold.white
    when /error/
      line.red
    when /\[PASSED\]/
      line.green
    when /\[PENDING\]/
      line.yellow
    when /^(.+?\.m)(:\d+:\s.+?\[FAILED\].+)/m
      if $1 == 'Unknown.m'
        line.red
      else
        (Pathname.new($1).relative_path_from(@source_root).to_s + $2).red
      end
    else
      line
    end
    self
  end
end

cmd = "xcodebuild -workspace #{workspace} -scheme #{scheme} -sdk iphonesimulator TEST_AFTER_BUILD=YES ONLY_ACTIVE_ARCH=NO build"
stdout = TestOutput.new(STDOUT, verbose, source_root)
stderr = TestOutput.new(STDERR, verbose, source_root)
status = Open4.spawn(cmd, :stdout => stdout, :stderr => stderr, :status => true)
exit status.exitstatus


SocialApp/Macros.h
/** Wraps the provided controller in a navigation controller. */
#define NAVIFY(controller) [[UINavigationController alloc] initWithRootViewController:controller]

/** Creates an NSURL out of the provided string. */
#define URLIFY(urlString) [NSURL URLWithString:urlString]

/** Formats a string */
#define F(string, args...) [NSString stringWithFormat:string, args]

/** Helper to create a quick UIAlertView with a title & message.  Contains a single OK button. */
#define ALERT(title, msg) [[[UIAlertView alloc] initWithTitle:title\
                                                      message:msg\
                                                     delegate:nil\
                                            cancelButtonTitle:@"OK"\
                                            otherButtonTitles:nil] show]

scripts/install_provisioning_profile.sh
#! /bin/sh

# Takes a provisioning profile and installs it into the system path.
# This requires poking inside the provisioning profile and searching for the
# UUID present at the end of the file.

set -e

if [[ -z "$1" ]]
then
  echo "USAGE: install_provisioning_profile PROVISIONING_FILE"
  exit 1
fi

if [[ ! -f $1 ]]
then
  echo "The file $1 does not exist"
  exit 1
fi

mobile_provisioning_file=$1
echo "Processing $mobile_provisioning_file"

uuid=`grep "UUID" -A1 -a "$mobile_provisioning_file" | grep -o "[-A-Z0-9]\{36\}"`
echo "uuid is $uuid"

if [[ -z "$uuid" ]]
then
  echo "UUID could not be found in $mobile_provisioning_file"
  exit 1
fi

destination="$HOME/Library/MobileDevice/Provisioning Profiles/$uuid.mobileprovision"
cp $mobile_provisioning_file "$destination"
echo "Installed to $destination"


SocialApp/TwitterFollowersViewController.h
//
//  TwitterFollowersViewController.h
//  SocialApp
//
//  Created by ben on 3/2/13.
//  Copyright (c) 2013 NSScreencast. All rights reserved.
//

#import <UIKit/UIKit.h>
#import <Accounts/Accounts.h>

/**
 `TwitterFollowersViewController` is a subclass of `UITableView` for showing the followers & avatars
 of a twitter account.
 
 ## Usage
 
 This view controller requires an `ACAccountStore` to be provided via the `accountStore` property
 before the controller is displayed.
 
 This view controller does not handle the `ACAccountStoreDidChangeNotification` however it may be
 dismissed if the presenting view controller listens for this notification.
 
 ## Limitations
 
 @warning This controller assumes there is only one twitter account in settings.  If multiple are present the last one added will be used.
 
 @warning Only the first 100 followers are listed.
*/
 
@interface TwitterFollowersViewController : UITableViewController

/** Sets the account store that is used to interact with twitter. */
@property (nonatomic, strong) ACAccountStore *accountStore;

@end


SocialApp/TwitterFollowersViewController.m
//
//  TwitterFollowersViewController.m
//  SocialApp
//
//  Created by ben on 3/2/13.
//  Copyright (c) 2013 NSScreencast. All rights reserved.
//

#import "TwitterFollowersViewController.h"
#import "TwitterFollowerCell.h"
#import "UIImageView+AFNetworking.h"
#import <Social/Social.h>

@interface TwitterFollowersViewController ()

@property (nonatomic, strong) NSDictionary *followerDictionary;

@end

@implementation TwitterFollowersViewController

- (id)initWithStyle:(UITableViewStyle)style {
    self = [super initWithStyle:style];
    if (self) {
    }
    return self;
}

- (UIBarButtonItem *)doneButton {
    return [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                                         target:self
                                                         action:@selector(onDone:)];
}

- (UIBarButtonItem *)tweetButton {
    return [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCompose
                                                         target:self
                                                         action:@selector(onCompose:)];
}

- (void)onDone:(id)sender {
    [self dismissViewControllerAnimated:YES
                             completion:nil];
}

- (void)onCompose:(id)sender {
    NSString *tweetText = @"I'm tweeting from iOS 6 😏";
    SLComposeViewController *tweetSheet = [SLComposeViewController composeViewControllerForServiceType:SLServiceTypeTwitter];
    [tweetSheet setInitialText:tweetText];
    [self presentViewController:tweetSheet
                       animated:YES
                     completion:^{
                         NSLog(@"Done tweeting");
                     }];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"Twitter Followers";
    self.navigationItem.rightBarButtonItem = [self doneButton];
    self.navigationItem.leftBarButtonItem = [self tweetButton];
    self.navigationController.navigationBar.tintColor = [UIColor colorWithRed:0 green:0.576 blue:0.752 alpha:1.000];
    
    [self.tableView registerClass:[TwitterFollowerCell class]
           forCellReuseIdentifier:@"Cell"];
    
    ACAccountStore *store = [[ACAccountStore alloc] init];
    ACAccountType *twitterAccountType = [store accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierTwitter];
    [store requestAccessToAccountsWithType:twitterAccountType
                                   options:nil
                                completion:^(BOOL granted, NSError *error) {
                                    if (granted) {
                                        [self fetchTwitterFollowers];
                                    } else {
                                        NSLog(@"Not granted: %@", [error localizedDescription]);
                                    }
                                }];
}

- (void)fetchTwitterFollowers {
    [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:YES];
    NSURL *url = [NSURL URLWithString:@"https://api.twitter.com/1.1/followers/list.json"];
    NSDictionary *params = @{@"screen_name": @"nsscreencast", @"skip_status": @"1", @"count":@"100"};
    ACAccountType *twitterAccountType = [self.accountStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierTwitter];
    NSArray *accounts = [self.accountStore accountsWithAccountType:twitterAccountType];
    SLRequest *request = [SLRequest requestForServiceType:SLServiceTypeTwitter
                                            requestMethod:SLRequestMethodGET
                                                      URL:url
                                               parameters:params];
    [request setAccount:[accounts lastObject]];
    
    [request performRequestWithHandler:^(NSData *responseData, NSHTTPURLResponse *urlResponse, NSError *error) {
        if (responseData) {
            if (urlResponse.statusCode >= 200 && urlResponse.statusCode < 300) {
                NSError *jsonError = nil;
                NSDictionary *followerData = [NSJSONSerialization JSONObjectWithData:responseData
                                                                             options:NSJSONReadingAllowFragments
                                                                               error:&jsonError];
                if (followerData) {
                    // NSLog(@"Response: %@", followerData);
                    self.followerDictionary = followerData;
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self.tableView reloadData];
                    });
                } else {
                    NSLog(@"JSON Parsing error: %@", jsonError);
                }
            } else {
                NSLog(@"Server returned HTTP %d", urlResponse.statusCode);
            }
        } else {
            NSLog(@"Something went wrong: %@", [error localizedDescription]);
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
        });
    }];
}

#pragma mark - Table view data source

- (NSArray *)followers {
    return self.followerDictionary[@"users"];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSUInteger count =[[self followers] count];
    NSLog(@"got %u users", count);
    return count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    
    NSDictionary *follower = [self followers][indexPath.row];
    cell.textLabel.text = follower[@"screen_name"];
    cell.detailTextLabel.text = follower[@"description"];
    
    NSURL *imageUrl = URLIFY(follower[@"profile_image_url"]);
    UIImage *defaultImage = [UIImage imageNamed:@"twitter_avatar.png"];
    cell.imageView.contentMode = UIViewContentModeCenter;
    [cell.imageView setImageWithURL:imageUrl placeholderImage:defaultImage];
    
    return cell;
}

@end


SocialApp/FacebookFriendsViewController.m
//
//  FacebookFriendsViewController.m
//  SocialApp
//
//  Created by ben on 3/2/13.
//  Copyright (c) 2013 NSScreencast. All rights reserved.
//

#import "FacebookFriendsViewController.h"
#import "UIImageView+AFNetworking.h"

@interface FacebookFriendsViewController ()

@property (nonatomic, strong) NSArray *friendsList;

@end

@implementation FacebookFriendsViewController

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (UIBarButtonItem *)postButton {
    return [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCompose
                                                         target:self
                                                         action:@selector(onPost:)];
}

- (void)onPost:(id)sender {
    SLComposeViewController *fbSheet = [SLComposeViewController composeViewControllerForServiceType:SLServiceTypeFacebook];
    [fbSheet setInitialText:@"Posting from iOS 6 😏"];
    [self presentViewController:fbSheet animated:YES
                     completion:nil];
}

- (UIBarButtonItem *)doneButton {
    return [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                                         target:self
                                                         action:@selector(onDone:)];
}

- (void)onDone:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];

    self.title = @"Facebook Friends";
    self.navigationController.navigationBar.tintColor = [UIColor colorWithRed:0.099 green:0.367 blue:0.654 alpha:1.000];
    self.navigationItem.leftBarButtonItem = [self postButton];
    self.navigationItem.rightBarButtonItem = [self doneButton];
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"Cell"];
    
    ACAccountType *fbAccountType = [self.accountStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierFacebook];
    NSDictionary *options = @{
                              ACFacebookAppIdKey: @"223964394414685",
                              ACFacebookAudienceKey : ACFacebookAudienceFriends,
                              ACFacebookPermissionsKey : @[@"email", @"read_friendlists"]
                            };
    [self.accountStore requestAccessToAccountsWithType:fbAccountType
                                               options:options
                                            completion:^(BOOL granted, NSError *error) {
                                                if (granted) {
                                                    NSLog(@"Granted!");
                                                    ACAccount *account = [[self.accountStore accountsWithAccountType:fbAccountType] lastObject];
                                                    [self fetchFacebookFriendsFor:account];
                                                } else {
                                                    NSLog(@"Not granted: %@", error);
                                                    [[[UIAlertView alloc] initWithTitle:@"Facebook Permissions Error"
                                                                                message:[error localizedDescription]
                                                                               delegate:nil
                                                                      cancelButtonTitle:@"OK"
                                                                      otherButtonTitles:nil] show];
                                                }
                                            }];
}

- (void)fetchFacebookFriendsFor:(ACAccount *)facebookAccount {
    SLRequest *friendsRequest = [SLRequest requestForServiceType:SLServiceTypeFacebook
                                                   requestMethod:SLRequestMethodGET
                                                             URL:URLIFY(@"https://graph.facebook.com/me/friends")
                                                      parameters:nil];
    friendsRequest.account = facebookAccount;
    [friendsRequest performRequestWithHandler:^(NSData *responseData, NSHTTPURLResponse *urlResponse, NSError *error) {
        if (responseData) {
            NSLog(@"Got a response: %@", [[NSString alloc] initWithData:responseData
                                                               encoding:NSUTF8StringEncoding]);
            if (urlResponse.statusCode >= 200 && urlResponse.statusCode < 300) {
                NSError *jsonError = nil;
                NSDictionary *friendsListData = [NSJSONSerialization JSONObjectWithData:responseData
                                                                                options:NSJSONReadingAllowFragments
                                                                                  error:&jsonError];
                if (jsonError) {
                    NSLog(@"Error parsing friends list: %@", jsonError);
                } else {
                    self.friendsList = friendsListData[@"data"];
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self.tableView reloadData];
                    });
                }
            }
        } else {
            NSLog(@"ERROR Connecting");
        }
    }];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.friendsList.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    
    NSDictionary *friendData = self.friendsList[indexPath.row];
    cell.textLabel.text = friendData[@"name"];
    
    UIImage *defaultAvatar = [UIImage imageNamed:@"facebook_avatar.png"];
    NSURL *friendImageUrl = URLIFY(F(@"https://graph.facebook.com/%@/picture", friendData[@"id"]));
    cell.imageView.contentMode = UIViewContentModeCenter;   
    [cell.imageView setImageWithURL:friendImageUrl placeholderImage:defaultAvatar];
    
    return cell;
}

@end
